Move i18n key specification from UI Schema options to ControlElement

lucas-koehler opened the pull request:
* The i18n base key can no longer be specified as part of the UI Schema `options`
* Add optional `i18n` property to `ControlElement`
* Adapt i18n key resolution
* Add explicit tests for i18n key extraction from UI Schema and Json Schema
* Add `Internationalizable` interface and `isInternationalized` type guard to clearly define internationalizable UISchema elements.

Fix #1942

sdirix commented on packages/core/src/i18n/i18nUtil.ts:
```diff
@@ -8,7 +8,10 @@ export const getI18nKeyPrefixBySchema = (
   uischema: UISchemaElement | undefined
 ): string | undefined => {
-  return uischema?.options?.i18n ?? schema?.i18n ?? undefined;
+  if (uischema && isControlElement(uischema) && uischema.i18n) {
```
At the moment we only support it for `ControlElement`s but for this implementation here I would generalize this a bit. For example via an interface `Internationalizable` which is extended by `ControlElement`.

The function would then work unchanged for all UI Schema elements of the future instead of only for `Control`s.

sdirix commented on packages/core/src/models/uischema.ts:
```diff
@@ -213,6 +213,11 @@ export interface ControlElement extends UISchemaElement, Scopable {
   label?: string | boolean | LabelDescription;
+  /**
+   * The i18n key for the control. It is used to identify the string that needs to be translated.
+   * It is suffixed with `.label`, `.description` and `.error.<keyword>` to derive the corresponding message keys for the control.
```
There can be many more suffixes depending on each control. I would only give them as examples.

sdirix commented on packages/core/src/models/uischema.ts:
```diff
@@ -244,6 +249,9 @@ export interface Categorization extends UISchemaElement {
   elements: (Category | Categorization)[];
 }
 
+export const isControlElement = (element: UISchemaElement): element is ControlElement =>
```
We already have a `isControl` type guard.

lucas-koehler commented:
@sdirix Thanks for the review. I rebased the changes onto master and added a second commit that provides a new `Internationalizable` interface and a typeguard checking whether its `i18n` property is set.

sdirix commented on packages/core/src/models/uischema.ts:
```diff
@@ -244,6 +253,10 @@ export interface Categorization extends UISchemaElement {
   elements: (Category | Categorization)[];
 }
 
+export const isInternationalizable = (element: unknown): element is Required<Internationalizable> => {
```
I would change the name a bit to better fit the interface name, i.e. `..able` -> optional prop and `...ed` -> required prop.

```suggestion
export const isInternationalized = (element: unknown): element is Required<Internationalizable> => {
```

lucas-koehler commented:
@sdirix Thanks for the suggestion, that indeed fits better :) I adapted the commit to name the new type guard `isInternationalized`.
